ircd/msgbuf.c

```diff
@@ -335,6 +388,34 @@ msgbuf_unparse_prefix(char *buf, size_t *buflen, const struct MsgBuf *msgbuf, un
 	return used;
 }
 
+int
+msgbuf_unparse_para(char *buf, size_t buflen, const struct MsgBuf *msgbuf)
+{
+	size_t orig_len = strlen(buf);
+	size_t new_len = orig_len;
+	int ret;
+
+	for (size_t i = 0; i < msgbuf->n_para; i++) {
+		const char *fmt;
+
+		if (i == msgbuf->n_para - 1 && (msgbuf->preserve_trailing || strchr(msgbuf->para[i], ' ') != NULL || *msgbuf->para[i] == ':')) {
+			fmt = i == 0 ? ":%s" : " :%s";
+		} else {
+			fmt = i == 0 ? "%s" : " %s";
+		}
+
+		/* rb_snprintf_append returns the complete length of the final string, not just the number of bytes it wrote */
+		ret = rb_snprintf_append(buf, buflen, fmt, msgbuf->para[i]);
```

Comment:
Does this not have the potential for adding a colon twice? If the parameter begins with `:` you're setting `fmt` to `" :%s"`, and then here you're using that as the format string but then also passing the same parameter again. If the parameter is `:foo` this will then result in appending ` ::foo`.

Reply:
Ah, I didn't realise it was skipped over, and so a beginning colon was data (that needs to be preserved), not framing.